src/hooks/coreSDK/useInitSafeCoreSDK.ts

```diff
@@ -1,18 +1,24 @@
 import { useEffect, useState } from 'react'
-import { useCurrentChain } from '@/hooks/useChains'
+import useChainId from '@/hooks/useChainId'
 import useWallet from '../wallets/useWallet'
 import useSafeInfo from '@/hooks/useSafeInfo'
 import { initSafeSDK, setSafeSDK } from '@/hooks/coreSDK/safeCoreSDK'
 
 export const useInitSafeCoreSDK = (): Error | null => {
-  const chain = useCurrentChain()
+  const chainId = useChainId()
   const wallet = useWallet()
   const { safe, safeLoaded } = useSafeInfo()
 
   const [error, setError] = useState<Error | null>(null)
 
   useEffect(() => {
-    if (!safeLoaded || !chain || !wallet || chain.chainId !== wallet.chainId) {
+    if (
+      !safeLoaded ||
+      !wallet?.provider ||
+      !wallet.chainId ||
+      chainId !== wallet.chainId ||
+      safe.chainId !== wallet.chainId
+    ) {
       return
```

Comment:
Shouldn't we set the SDK to null or undefined here?

Reply:
I agree it would make sense to do that. There is one case where it doesn't work in safe apps:

1. Open a safe app on network A e.g. Drain Account
2. Switch wallet network to B
3. Press Transfer everything
4. It shows `Your wallet is connected to the wrong chain.` which is correct
5. Switch wallet network to A
6. It shows `The Safe SDK could not be initialized. Please be aware that we only support v1.1.1 Safes and up.` which is not correct

It seems that the `getSafeSDK` call inside the modal happens sooner than the `initSafeSDK` call that we do in Step 5 so it returns `undefined` and fails.

One possible solution I can think of is to use the `useSafeSDK` hook and pass it in the `useAsync` call inside `ReviewSafeAppsTx` to create the tx but it would probably still flash the error once.

Reply:
`getSafeSDK` should never be used in React components anyway, we should use a hook.

Reply:
I guess there's no harm in keeping the SDK instance stale. The submit button is disabled. And it will be re-initialized when the user switches to the right chain.

Reply:
Thinking a bit more about this we should make sure that it uses the re-initialized SDK in the case I described and not just "masks" the error.

Reply:
> I guess there's no harm in keeping the SDK instance stale. The submit button is disabled. And it will be re-initialized when the user switches to the right chain.

There are 2 issues that @francovenica found:

1. Its possible that the modal displays a wrong nonce because even with a stale SDK instance, the createTx call succeeds if both safes have the same owner.
2. Its possible to open the transaction modal from within a safe app even if the wallet is on the wrong network. This is something we can't control.

I propose that we reset the SDK instance as discussed which would solve the first issue. For the second issue I can think of multiple solutions:

1. Use `useSafeSDK` and pass the instance to its txSender function. We can limit this to the safe app modal because everywhere else the user can't open the modal if they are on the wrong network
2. Add logic to the `AppFrame` to close the modal if there is no SDK instance. This has the downside that the user is still able to "click" the button but nothing would happen in the UI.

Reply:
> Its possible to open the transaction modal from within a safe app even if the wallet is on the wrong network. This is something we can't control.

But they won't be able to submit it, right? Unless they switch the chain to the right one. At which point it should re-initialize the SDK.

Reply:
Exactly, this is what currently happens but since the rendered component that is visible inside the modal doesn't have a dependency to the core SDK it will not rerender and re-run the logic inside the `useAsync` where it accesses the getter.

Since it is possible to reset the Core SDK from anywhere inside the app and at any time (via the wallet) we can't reliably use `getSafeSDK` but should instead always rely on `useSafeSDK` to detect changes. Right now I only implemented this for certain calls with a fallback to `getSafeSDK` but imo we should adjust it to always pass the SDK as a dependency to those functions inside `txSender`